File: sdk/php/tests/OutcomeTest.php
<?php

declare(strict_types=1);

namespace Shojiku\Tests;

use PHPUnit\Framework\TestCase;
use Shojiku\Exception\UsageException;
use Shojiku\Origin;
use Shojiku\Outcome;
use Shojiku\Report;
use Shojiku\Step;

/**
 * The two failure levels, kept apart.
 *
 * `class: usage` is the CALLER's mistake and throws; everything a DOCUMENT can
 * do wrong comes back as a failed result with the engine's diagnostics
 * attached. An SDK that raised for the second class would have broken the
 * contract, not chosen an idiom.
 *
 * These build the envelope directly for the cases the real engine cannot be
 * asked to produce on demand — every OTHER test in the suite drives the same
 * code through the real binary.
 */
final class OutcomeTest extends TestCase
{
    use EngineFixtures;

    public function testACallerErrorThrowsRatherThanReturningAFailedResult(): void
    {
        $report = Report::parse((string) json_encode([
            'ok' => false,
            'diagnostics' => ['items' => []],
            'failure' => [
                'class' => 'usage',
                'step' => 'sign',
                'kind' => 'passphrase_variable',
                'message' => 'the environment variable `X` is not set',
            ],
        ]));

        self::assertTrue($report->isUsage());
        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/the engine refused the call.*is not set/');
        Outcome::verdict($report);
    }

    public function testARefusedDocumentIsAFailedResultCarryingItsDiagnostics(): void
    {
        $report = Report::parse((string) json_encode([
            'ok' => false,
            'diagnostics' => ['items' => [['severity' => 'error', 'code' => 'image_source_missing']]],
            'failure' => [
                'class' => 'document',
                'step' => 'render',
                'kind' => 'document',
                'message' => 'validation failed with errors',
            ],
        ]));

        $result = Outcome::document($report, '', Step::Generate, $this->client(), Origin::Rendered);

        self::assertTrue($result->failed());
        self::assertSame(Step::Generate, $result->failure()?->step());
        self::assertSame('document', $result->failure()?->kind());
        self::assertCount(1, $result->errors());
    }

    public function testASuccessWithDiagnosticsIsStillASuccess(): void
    {
        $report = Report::parse((string) json_encode([
            'ok' => true,
            'diagnostics' => ['items' => [['severity' => 'warning', 'code' => 'text_overflow']]],
            'pageCount' => 3,
        ]));

        $result = Outcome::document($report, '%PDF-x', Step::Generate, $this->client(), Origin::Source);

        self::assertTrue($result->success());
        self::assertCount(1, $result->warnings());
        self::assertSame(3, $result->artifact()?->pageCount());
        self::assertSame(Origin::Source, $result->artifact()?->origin());
    }

    public function testAFailureBlockThatIsMalformedReportsUnknownRatherThanGuessing(): void
    {
        $report = Report::parse('{"ok":false,"failure":{"class":"document"}}');

        $result = Outcome::verdict($report);

        self::assertTrue($result->failed());
        self::assertSame('unknown', $result->failure()?->kind());
        self::assertSame('', $result->failure()?->message());
    }

    public function testAFailureBlockThatIsNotAnObjectIsTreatedAsAbsent(): void
    {
        // `ok` already says which way the operation went; inventing a kind an
        // SDK branches on would be worse than reporting none.
        $report = Report::parse('{"ok":true,"failure":"nope"}');

        self::assertFalse($report->isUsage());
        self::assertTrue(Outcome::verdict($report)->success());
    }

    public function testAVerdictWithNoVerificationBlockHasNoReport(): void
    {
        $report = Report::parse('{"ok":false,"failure":{"class":"document","kind":"verify","message":"none"}}');

        $result = Outcome::verdict($report);

        self::assertTrue($result->failed());
        self::assertNull($result->report());
    }

    public function testAPageCountThatIsNotAnIntegerIsAbsentRatherThanCast(): void
    {
        $report = Report::parse('{"ok":true,"pageCount":"three"}');

        self::assertNull($report->pageCount);
    }
}

File: sdk/java/src/test/java/jp/kengos/shojiku/OutcomeTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import org.junit.jupiter.api.Test;

/**
 * The two levels of failure the C surface defines, and keeping them apart.
 *
 * <p>A non-zero status is the CALLER's mistake and throws; everything a DOCUMENT can do wrong comes
 * back as a failed result with the engine's diagnostics attached. An SDK that threw on the second
 * class would have broken the contract, not chosen an idiom.
 */
class OutcomeTest extends Fixtures {

  private static Snapshot snapshot(
      int status, boolean success, byte[] pdf, String json, String diagnostics, String error) {
    return new Snapshot(status, success, pdf, json, diagnostics, error);
  }

  private static Snapshot ok() {
    return snapshot(0, true, new byte[0], "", "", "");
  }

  @Test
  void aNonZeroStatusIsProgrammerMisuse() {
    UsageException error =
        assertThrows(
            UsageException.class,
            () -> Outcome.guard(snapshot(3, false, new byte[0], "", "", "{\"kind\":\"bad\"}")));

    assertTrue(error.getMessage().contains("status 3"));
  }

  @Test
  void aZeroStatusPassesThrough() {
    Outcome.guard(ok());
  }

  @Test
  void aDocumentFailureIsAFailedResultRatherThanAnException() {
    Result<DocumentArtifact> result =
        Outcome.document(
            snapshot(
                0,
                false,
                new byte[0],
                "",
                "{\"items\":[{\"severity\":\"error\",\"message\":\"boom\"}]}",
                "{\"kind\":\"render_failed\",\"message\":\"no\"}"),
            Step.GENERATE,
            null,
            Origin.RENDERED);

    assertTrue(result.failed());
    assertEquals("render_failed", result.failure().kind());
    assertEquals(1, result.errors().size());
  }

  @Test
  void aSuccessCarriesItsDiagnosticsToo() {
    // A render that WORKED can still have warned, and a caller that only looks at
    // failures never sees it.
    Result<DocumentArtifact> result =
        Outcome.document(
            snapshot(
                0,
                true,
                new byte[] {1, 2, 3},
                "{\"pageCount\":2}",
                "{\"items\":[{\"severity\":\"warning\",\"message\":\"cramped\"}]}",
                ""),
            Step.GENERATE,
            null,
            Origin.SOURCE);

    assertTrue(result.success());
    assertEquals(2, result.artifact().pageCount());
    assertEquals(Origin.SOURCE, result.artifact().origin());
    assertEquals(1, result.warnings().size());
  }

  @Test
  void aPageCountIsAbsentRatherThanZeroWhenNothingLaidAnythingOut() {
    // Signing appends a revision to bytes it never measured, and the surface
    // returns no JSON payload for it at all.
    Result<DocumentArtifact> signed =
        Outcome.document(
            snapshot(0, true, new byte[] {1}, "", "", ""), Step.SIGN, null, Origin.RENDERED);

    assertNull(signed.artifact().pageCount());
  }

  @Test
  void aJsonPayloadWithoutAPageCountIsAbsentToo() {
    Result<DocumentArtifact> result =
        Outcome.document(
            snapshot(0, true, new byte[] {1}, "{\"other\":1}", "", ""),
            Step.GENERATE,
            null,
            Origin.RENDERED);

    assertNull(result.artifact().pageCount());
  }

  @Test
  void aVerdictParsesItsReportBeforeReadingTheVerdict() {
    // Because the report rides a FAILED verify too — that is the whole point of
    // carrying notChecked.
    Result<VerificationReport> result =
        Outcome.verdict(
            snapshot(
                0,
                false,
                new byte[0],
                "{\"valid\":false,\"notChecked\":[\"revocation\"]}",
                "",
                "{\"kind\":\"signature\",\"message\":\"digest mismatch\"}"));

    assertTrue(result.failed());
    assertNotNull(result.report());
    assertEquals(List.of("revocation"), result.report().notChecked());
    assertEquals(Step.VERIFY, result.failure().step());
    assertEquals("signature", result.failure().kind());
  }

  @Test
  void aVerdictWithNoPayloadCarriesNoReportAndThatAbsenceIsData() {
    // A different fact from an empty report: there was nothing to evaluate.
    Result<VerificationReport> result = Outcome.verdict(ok());

    assertTrue(result.success());
    assertNull(result.report());
  }

  @Test
  void aVerdictCarriesDiagnosticsOnBothPaths() {
    // Same reason they ride a render: whatever the engine noticed belongs to the
    // caller, and an operation that drops them makes its result mean something
    // different from every other operation's.
    String diagnostics = "{\"items\":[{\"severity\":\"warning\",\"message\":\"noted\"}]}";

    assertEquals(
        1,
        Outcome.verdict(snapshot(0, true, new byte[0], "{\"valid\":true}", diagnostics, ""))
            .warnings()
            .size());
    assertEquals(
        1, Outcome.verdict(snapshot(0, false, new byte[0], "", diagnostics, "")).warnings().size());
  }

  @Test
  void theGuardRunsBeforeEitherKindOfResultIsBuilt() {
    // A caller error is not a document outcome, on any path.
    Snapshot refused = snapshot(1, false, new byte[0], "", "", "");

    assertThrows(
        UsageException.class, () -> Outcome.document(refused, Step.GENERATE, null, Origin.LOADED));
    assertThrows(UsageException.class, () -> Outcome.verdict(refused));
  }
}
